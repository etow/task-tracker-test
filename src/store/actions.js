import taskRepository from '../repositories/taskRepository';
import taskCategoryRepository from '../repositories/taskCategoryRepository';
import {
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASKS,
  ADD_TASK,
  DELETE_TASK,
  SET_TASK_TO_EDIT,
  SET_TASK,
  FETCH_TASKS_CATEGORIES,
  FETCH_TASKS,
  CREATE_TASK,
  UPDATE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
  UPDATE_TASKS,
} from './constants';

export default {
  [FETCH_TASKS_CATEGORIES]: ({ commit }) => {
    return taskCategoryRepository.get()
      .then((response) => {
        commit(SET_TASKS_CATEGORIES, response.data);
        commit(SET_TASKS_CATEGORIES_DATA, response.data);
      });
  },
  [FETCH_TASKS]: ({ commit }) => {
    return taskRepository.get()
      .then((response) => commit(SET_TASKS, response.data));
  },
  [CREATE_TASK]: ({ commit, state }, newTask) => {
    commit(BACKUP_STATE, state);
    const task = {
      ...newTask,
      id: Math.random().toString(36).substr(2, 6),
      order: state.tasks[newTask.category].length,
    };
    commit(ADD_TASK, task);
    return taskRepository.create(task);
  },
  [DELETE_TASK]: ({ commit, state }, {task, category}) => {
    commit(BACKUP_STATE, state);
    commit(DELETE_TASK, {task, category});
    commit(SET_TASK_TO_EDIT, {});
    return taskRepository.delete(task.id);
  },
  [UPDATE_TASK]: ({ commit, state }, {task, prevCategory}) => {
    commit(BACKUP_STATE, state);
    if (prevCategory && task.category !== prevCategory) {
      commit(DELETE_TASK, { task, category: prevCategory });
    }
    const targetIndex = state.tasks[task.category].findIndex((aTask) => aTask.id === task.id);
    if (targetIndex >= 0) {
      commit(SET_TASK, { task, targetIndex });
    } else {
      const order = state.tasks[task.category].length;
      commit(ADD_TASK, { ...task, order });
    }
    commit(SET_TASK_TO_EDIT, {})
    return taskRepository.update(task);
  },
  [UPDATE_TASKS]: ({ commit, state }, { tasks, category }) => {
    commit(BACKUP_STATE, state);
    tasks.forEach((task, index) => {
      task.category = category;
      task.order = index;
    });
    commit(UPDATE_TASKS, { tasks, category });

    return taskRepository.update(tasks);
  },
  [SET_TASK_TO_EDIT]: ({ commit, state}, task) => {
    commit(BACKUP_STATE, state);
    commit(SET_TASK_TO_EDIT, task);
  },
  [ROLLBACK_STATE]: ({ commit }) => {
    commit(ROLLBACK_STATE);
  },
};
