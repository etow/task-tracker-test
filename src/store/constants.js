

/* mutations and actions */
export const ADD_TASK = 'ADD_TASK';
export const SET_TASKS = 'SET_TASKS';
export const SET_TASKS_CATEGORIES = 'SET_TASKS_CATEGORIES';
export const SET_TASKS_CATEGORIES_DATA = 'SET_TASKS_CATEGORIES_DATA';
export const SET_TASK = 'SET_TASK';
export const SET_TASK_TO_EDIT = 'SET_TASK_TO_EDIT';
export const DELETE_TASK = 'DELETE_TASK';
export const FETCH_TASKS_CATEGORIES = 'FETCH_TASKS_CATEGORIES';
export const FETCH_TASKS = 'FETCH_TASKS';
export const CREATE_TASK = 'CREATE_TASK';
export const UPDATE_TASK = 'UPDATE_TASK';
export const ROLLBACK_STATE = 'ROLLBACK_STATE';
export const BACKUP_STATE = 'BACKUP_STATE';
export const UPDATE_TASKS = 'UPDATE_TASKS';
export const UPDATE_TASK_ACTIVITY = 'UPDATE_TASK_ACTIVITY';
