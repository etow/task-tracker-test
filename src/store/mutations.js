import {
  ADD_TASK,
  SET_TASKS,
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASK,
  SET_TASK_TO_EDIT,
  DELETE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
} from './constants';

export default {
  [ADD_TASK]: (state, task) => {
    state.tasks[task.category].push(task);
  },
  [SET_TASKS]: (state, tasks) => {
    Object.keys(state.tasks)
      .forEach((category) => {
        const tasksByCategory = tasks.filter((task) => task.category === category);
        state.tasks[category] = tasksByCategory.sort((a, b) => a.order - b.order);
      });
  },
  [SET_TASKS_CATEGORIES]: (state, categories) => {
    categories.forEach((cat) => state.tasks[cat.name] = []);
  },
  [SET_TASKS_CATEGORIES_DATA]: (state, categories) => {
    categories.forEach((cat) => state.categories[cat.name] = cat);
  },
  [SET_TASK]: (state, { task, targetIndex }) => {
    state.tasks[task.category][targetIndex] = task;
  },
  [SET_TASK_TO_EDIT]: (state, task) => {
    state.taskToEdit = task;
  },
  [DELETE_TASK]: (state, {task, category}) => {
    state.tasks[category] = state.tasks[category].filter((aTask) => aTask.id !== task.id);
  },
  [ROLLBACK_STATE]: (state) => {
    Object.assign(state, JSON.parse(JSON.stringify(state.prevState)));
  },
  [BACKUP_STATE]: (state, currentState) => {
    state.prevState = JSON.parse(JSON.stringify(currentState));
  }
}
