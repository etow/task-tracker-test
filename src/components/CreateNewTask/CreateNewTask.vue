<template>
  <div class="create-new-task">
    <ElInput
      v-show="editMode"
      ref="textInputRef"
      v-model="task.name"
      type="text"
      @blur="handleBlur"
      @keypress="handleKeyPress" />
    <div v-show="!editMode" style="text-align: center;">
      <ElButton round @click="enableEditMode">
        +
      </ElButton>
    </div>
  </div>
</template>

<script>
import { CREATE_TASK, ROLLBACK_STATE } from '../../store/constants';
import { ElMessage, ElInput, ElButton } from 'element-plus';

export default {
  props: {
    category: {
      type: String,
      required: true,
    },
  },
  components: {
    ElInput,
    ElButton,
  },
  data() {
    return {
      editMode: false,
      task: { name: '', category: this.category },
      taskCreated: false,
    };
  },
  methods: {
    createNewTask() {
      this.editMode = false;
      // prevent creating task twice (enter + blur);
      if (this.taskCreated) return;
      // if name is empty, do not create task
      if (trimSpaces(this.task.name) !== '') {
        this.taskCreated = true;
        this.$store.dispatch(CREATE_TASK, this.task)
          .then(() => this.taskCreated = false)
          .catch((err) => {
            console.error(err.message);
            ElMessage({
              message: 'Something went wrong while trying to create a new task, please try again',
              type: 'error',
              duration: 5000,
            });
            this.$store.dispatch(ROLLBACK_STATE);
          })
          .finally(() => this.taskCreated = false);
      }
    },
     handleBlur() {
       this.createNewTask();
     },
     handleKeyPress(e) {
       if (e.key === 'Enter') this.createNewTask();
     },
     enableEditMode() {
      this.editMode = true;
      this.task = { name: '', category: this.category };
      this.$refs.textInputRef.focus();
     },
  },
};

const trimSpaces = (str) => {
  let trimmed = str.trim();
  trimmed.replace(new RegExp(String.fromCharCode(160), 'g'), '');
  trimmed = trimmed.replace(/\s+/g, ' ');
  return trimmed;
};
</script>
<style lang="scss" scoped>
  ::v-deep(.el-input__inner) {
    border-radius: 15px;
    box-shadow: rgba(0, 0, 0, 0.02) 0px 0px 0px 1px, rgba(0, 0, 0, 0.05) 0px 1px 2px 0px, rgba(0, 0, 0, 0.05) 0px 2px 8px 0px;
    height: 58px;
    line-height: 58px;
  }
</style>
