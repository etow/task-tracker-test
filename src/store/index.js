import { createStore } from 'vuex';

import mutations from './mutations';
import actions from './actions';

export default createStore({
  state: {
    tasks: {},
    categories: {},
    taskToEdit: {},
    prevState: {},
  },
  mutations,
  actions,
});
