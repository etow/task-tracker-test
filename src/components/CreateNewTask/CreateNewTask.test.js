import { flushPromises, shallowMount } from '@vue/test-utils'
import { ElMessage } from 'element-plus';
import CreateNewTask from './CreateNewTask.vue'
import { createStore } from 'vuex'
import { CREATE_TASK, ROLLBACK_STATE } from '../../store/constants';

jest.mock('element-plus', () => ({
  ElMessage: jest.fn(),
}));

describe('CreateNewTask.vue', () => {
  it('It renders correctly', () => {
    const wrapper = createWrapper();
    expect(wrapper.html()).toMatchSnapshot();
  })

  it('It renders edit mode', async () => {
    // mock focus
    const inputFocus = jest.fn();

    // create wrapper
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.stubs.ElInput.methods.focus = inputFocus;
      return dataToMount;
    });

    // check click on button and focus on input.
    const button = await wrapper.find('button');
    const input = await wrapper.find('input');

    // check input is hidden
    expect(input.wrapperElement.style.display).toEqual('none');

    button.trigger('click');
    await wrapper.vm.$nextTick();

    // check input is visible after click
    expect(inputFocus).toHaveBeenCalled();
    expect(input.wrapperElement.style.display).toEqual('');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('It creates a new task on blur', async () => {
    // mock store
    const [store, { actions }] = mockStore();

    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    // mock values
    wrapper.vm.task.name = 'Pass Test';
    const createTaskMethod = jest.spyOn(wrapper.vm, 'createNewTask');

    // check input blur
    const input = await wrapper.find('input');
    input.trigger('blur');

    expect(createTaskMethod).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.taskCreated).toBe(true);
    expect(actions[CREATE_TASK]).toHaveBeenCalledTimes(1);

    await flushPromises();

    // resets state
    expect(wrapper.vm.taskCreated).toBe(false);
    expect(wrapper.vm.editMode).toBe(false);
  });

  it('It creates a new task on keypress = ENTER', async () => {

    // mock store
    const [store, { actions }] = mockStore();

    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    // mock values
    wrapper.vm.task.name = 'Pass Test';
    const createTaskMethod = jest.spyOn(wrapper.vm, 'createNewTask');


    // check input keypress
    const input = await wrapper.find('input');
    input.trigger('keypress', { key: 'Enter' });

    expect(createTaskMethod).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.taskCreated).toBe(true);
    expect(actions[CREATE_TASK]).toHaveBeenCalledTimes(1);

    await flushPromises();

    // resets state
    expect(wrapper.vm.taskCreated).toBe(false);
    expect(wrapper.vm.editMode).toBe(false);
  });

  it('It cancel create new task if method has already been executed', async () => {

    // mock store
    const [store, { actions }] = mockStore();

    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    // mock values
    wrapper.vm.task.name = 'Pass Test';
    const createTaskMethod = jest.spyOn(wrapper.vm, 'createNewTask');

    // trigger consecutive events
    const input = await wrapper.find('input');
    input.trigger('blur');
    input.trigger('keypress', { key: 'Enter' });

    expect(createTaskMethod).toHaveBeenCalledTimes(2);
    expect(actions[CREATE_TASK]).toHaveBeenCalledTimes(1);

    await flushPromises();

    // resets state
    expect(wrapper.vm.taskCreated).toBe(false);
    expect(wrapper.vm.editMode).toBe(false);
  });

  it('It cancel create new task if task name is empty', async () => {

    // mock store
    const [store, { actions }] = mockStore();

    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    // mock values
    const createTaskMethod = jest.spyOn(wrapper.vm, 'createNewTask');

    // trigger consecutive events
    const input = await wrapper.find('input');
    input.trigger('blur');

    expect(createTaskMethod).toHaveBeenCalledTimes(1);
    expect(actions[CREATE_TASK]).toHaveBeenCalledTimes(0);

    await flushPromises();

    // resets state
    expect(wrapper.vm.taskCreated).toBe(false);
    expect(wrapper.vm.editMode).toBe(false);
  });

  it('It reset states when switching modes', async () => {

    // mock store
    const [store] = mockStore();

    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    // initial state
    expect(wrapper.vm.taskCreated).toBe(false);
    expect(wrapper.vm.task).toEqual({ name: '', category: 'Test' });

    // triger event
    const button = await wrapper.find('button');
    const input = await wrapper.find('input');

    button.trigger('click');
    await flushPromises();

    expect(wrapper.vm.editMode).toBe(true);
    expect(wrapper.vm.task).toEqual({ name: '', category: 'Test' });

    wrapper.vm.task.name = 'Pass test again';
    input.trigger('blur');

    await flushPromises();

    // resets state
    expect(wrapper.vm.taskCreated).toBe(false);
    expect(wrapper.vm.editMode).toBe(false);

    // enter edit mode again
    button.trigger('click');

    await flushPromises();

    expect(wrapper.vm.editMode).toBe(true);
    expect(wrapper.vm.task).toEqual({ name: '', category: 'Test' });
  });

  it('it handles error when creating a task', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});

    const [store, { actions }] = mockStore({ create_new_task: jest.fn(() => Promise.reject('error'))});
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    // mock values
    wrapper.vm.task.name = 'Pass Test';
    wrapper.vm.createNewTask();

    expect(actions[CREATE_TASK]).toHaveBeenCalled();
    await flushPromises();

    expect(console.error).toHaveBeenCalled();
    expect(ElMessage).toHaveBeenCalledWith({
      message: 'Something went wrong while trying to create a new task, please try again',
      type: 'error',
      duration: 5000,
    });
    expect(actions[ROLLBACK_STATE]).toHaveBeenCalled();
    console.error.mockRestore();
  });

})

// create wrapper helper
const createWrapper = (decorator = (n) => n) => {
  const dataToMount = {
    propsData: {
      category: 'Test',
    },
    global: {
      stubs: {
        ElButton: {
          template: '<button />',
        },
        ElInput: {
          template: '<input />',
          methods: {
            focus: jest.fn(),
          },
        },
      },
    },
  }

  decorator(dataToMount);
  return shallowMount(CreateNewTask, dataToMount);
};

// mock store helper
function mockStore({create_new_task, rollback_state} = {}) {
  const actions = {
    [CREATE_TASK]: create_new_task || jest.fn(() => Promise.resolve()),
    [ROLLBACK_STATE]: rollback_state || jest.fn(() => Promise.resolve()),
  };
  const store = createStore({
    actions,
  });

  return [store, { actions }];
}
