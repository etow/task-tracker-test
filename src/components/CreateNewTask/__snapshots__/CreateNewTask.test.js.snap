// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewTask.vue It renders correctly 1`] = `
<div class="create-new-task"><input modelvalue="" type="text" style="display: none;">
  <div style="text-align: center;"><button round=""></button></div>
</div>
`;

exports[`CreateNewTask.vue It renders edit mode 1`] = `
<div class="create-new-task"><input modelvalue="" type="text" style="">
  <div style="text-align: center; display: none;"><button round=""></button></div>
</div>
`;
